llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,466 @@
+
+  void getAnalysisUsage(AnalysisUsage &AU) const override {
+    AU.setPreservesCFG();
+    AU.addRequired<VirtRegMap>();
+    AU.addPreserved<VirtRegMap>();
+    AU.addRequired<SlotIndexes>();
+    AU.addPreserved<SlotIndexes>();
+    AU.addRequired<LiveIntervals>();
+    AU.addPreserved<LiveIntervals>();
+    AU.addRequired<LiveRegMatrix>();
+    AU.addPreserved<LiveRegMatrix>();
+    AU.addRequired<MachineLoopInfo>();
+    AU.addPreserved<MachineLoopInfo>();
+    MachineFunctionPass::getAnalysisUsage(AU);
+  }
+
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+
+private:
+};
+
+/// Find Free register of the same register Class type and
+// exclude the forbidden Physical register from the result
+
+static llvm::SmallVector<MCPhysReg, 16> getPriorityReplacementRegs(
+    const Register Reg,
+    const llvm::SmallVector<MCPhysReg, 16> &ForbiddenPhysRegs,
+    const MachineFunction &MF, const VirtRegMap &VRM, LiveRegMatrix &LRM,
+    const MachineRegisterInfo &MRI, const LiveIntervals &LIS,
+    const AIEBaseRegisterInfo &TRI);
+
+static bool
+replaceReg(const Register Reg,
+           const llvm::SmallVector<MCPhysReg, 16> &ForbiddenPhysRegs,
+           llvm::SmallVector<MCPhysReg, 16> &AlreadyReplaced,
+           const MachineFunction &MF, VirtRegMap &VRM, LiveRegMatrix &LRM,
+           const MachineRegisterInfo &MRI, const LiveIntervals &LIS,
+           const AIEBaseRegisterInfo &TRI);
+
+static llvm::SmallVector<const MachineBasicBlock *, 4>
+getLoopBBs(const MachineLoopInfo &MLI);
+static bool isWAWCandidate(const Register &Reg, const MachineInstr &MI,
+                           llvm::SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+                           llvm::SmallVector<MCPhysReg, 16> &UsedVirtRegs,
+                           VirtRegMap &VRM, MachineRegisterInfo &MRI);
+
+bool isCopyElimination(const MachineInstr &MI, const VirtRegMap &VRM);
+
+static llvm::SmallVector<MCPhysReg, 16>
+getBBForbiddenReplacements(const MachineBasicBlock *MBB, const VirtRegMap &VRM,
+                           const MachineRegisterInfo &MRI,
+                           const AIEBaseRegisterInfo *TRI);
+
+static bool
+isUndefReserved(const MachineOperand &Def, const MachineInstr &MI,
+                const MachineBasicBlock &MBB,
+                std::map<unsigned, llvm::SmallVector<const MachineInstr *, 4>>
+                    &UndefPartReuse,
+                const VirtRegMap &VRM);
+
+static bool isNotLastDef(const MachineOperand &Def,
+                         const MachineInstr &DefInstr,
+                         const MachineBasicBlock &MBB);
+
+static bool
+reuseSubRegs(llvm::SmallVector<const MachineInstr *, 4> &OrigInstr,
+             llvm::SmallVector<const MachineInstr *, 4> &OtherInstr);
+
+static bool containEqualSubRegDefs(const MachineInstr *InstrA,
+                                   const MachineInstr *InstrB);
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+  if (DisablePass)
```

Comment:
You can remove this after changing the pass creating place, or use:

```
  if (skipFunction(MF.getFunction()))
    return false;
```

To rely on function attribute (it can be independent of the optimization level) to decide about the optimization.

Reply:
Since i removed the enable option, this is alsomoved